# Section 00 - Getting Ready
1. Make sure you have already installed and loaded the following libraries:
```{r}
library(ggplot2)
library(data.table)
library(magrittr)
library(tidyr)
library(dplyr)
library(patchwork) # optional, makes plots nicer
```
2. Load the yeast data
```{r}
genotype <- fread("./extdata/eqtl/genotype.txt")
genotype <- melt(genotype, id.vars = "strain", variable.name = "marker",
value.name = "genotype")
growth <- fread("./extdata/eqtl/growth.txt")
growth <- melt(growth, id.vars = "strain", variable.name = "media",
value.name = "growth_rate")
marker <- fread("./extdata/eqtl/marker.txt")
```



# Section 01 - Permutation test of growth rate difference
1. The following code recreates the example shown in the lecture to test the association of the genotype
at marker 5211 with the growth rate difference in Maltose medium. Note that the code is written using
functions, meaning that it will work for any marker, not just marker 5211. Read it carefully to understand
what happens in each function. Then execute the code. The Lecture example and a description of the dataset
can be found here: https://gagneurlab.github.io/dataviz/resampling-stat.html#yeast-dataset.
# Plotting the growth rate difference
```{r}
getMaltoseDt = function(mrk) {
  growth_mrk <- merge(growth, genotype[marker %in% mrk, .(strain, genotype, marker)],
  by = 'strain', allow.cartesian = TRUE)
  growth_mrk[media == "YPMalt"]
}
# boxplot
plot_growth_one_mk <- function(mk){
  ggplot(getMaltoseDt(mk), aes(genotype, growth_rate)) +
  geom_boxplot() +
  labs(title = mk) + theme_bw(base_size = 16)
}
plot_growth_one_mk("mrk_5211")

# Function to calculate the difference of the medians of two genotypes
median_diff <- function(dt){
  dt[genotype == 'Wild isolate', median(growth_rate, na.rm=T)] -
  dt[genotype == 'Lab strain', median(growth_rate, na.rm=T)]
}

# Function to permute the table, plot the resulting histogram
# and compute a p-value
p_val_medians <- function(dt, N_permu = 1000){
  
  # It will return both a pvalue and plot a histogram of T_star
  T_ref <- median_diff(dt)
  T_star <- sapply(1:N_permu, function(x) {
    median_diff(dt[, genotype := sample(genotype)]) 
  })
  # Plot
  g <- ggplot(data = data.table(T_star = T_star), aes(T_star)) + geom_histogram() +
  geom_vline(aes(xintercept=T_ref, color="T_ref")) + xlim(-3,3)
  print(g) # Needed to render plot inside function call
  # Compute and return the p value
  # First compute each tail seperately
  p_val_right <- (sum(T_star >= T_ref) + 1) / (N_permu + 1)
  p_val_left <- (sum(T_star <= T_ref) + 1) / (N_permu + 1)
  # Then combine the above to obtain the double sided p-value.
  p_val <- 2 * min(p_val_right, p_val_left)
  return(p_val)
}
# Calling the function:
p_val_medians(getMaltoseDt("mrk_5211"))
```
2. Using the code above, plot and test whether markers 1653 and 5091 associate with growth. Interpret your
results.
```{r}
plot_growth_one_mk("mrk_1653")
p_val_medians(getMaltoseDt("mrk_1653"))

plot_growth_one_mk("mrk_5091")
p_val_medians(getMaltoseDt("mrk_5091"))
```
mrk_1653 not statistically associated with growth
mrk_5091 statistically associated with growth



# Section 02 - Permutation test of marker association
1. We just concluded that both markers 5211 and 5091 are significantly associated with growth. However,
this could be confounded. A common source of confounding in genomics is due to “linkage”, which describes
the phenomenon of markers being inherited together. A biological explanation for linkage is provided
here: https://www.khanacademy.org/science/biology/classical- genetics/chromosomal- basis- of -
genetics/a/linkage-mapping
To investigate the issue of linkage in our dataset, test if marker 5091 significantly associates with marker
5211. Define a null hypothesis, a statistic and use permutation testing to answer the question. Strengthen
your answer with a relevant plot.
Hint: start from the table below and think about how this can be permuted.
```{r}
mks_geno <- dcast(genotype[marker %in% c("mrk_5091", "mrk_5211")], ...~marker)
mks_geno
```
```{r}
get_links <- function(dt){
  dt[mrk_5091==mrk_5211, .N]
}
get_links(mks_geno)


p_val_links <- function(dt, N_permu=1000) {
  T_ref <- get_links(dt)
  T_star <- sapply(1:N_permu, function(x) {
    get_links(dt[, .(mrk_5211=sample(mrk_5211), mrk_5091=sample(mrk_5091))])
  })
  
  g <- ggplot(data.table(T_star=T_star), aes(T_star)) +
    geom_histogram() +
    geom_vline(aes(xintercept=T_ref, color="T_ref"))
  
  print(g)
  
  p_val_right <- (sum(T_star >= T_ref) + 1) / (N_permu + 1)
  p_val_left  <- (sum(T_star <= T_ref) + 1) / (N_permu + 1)
  
  p_val <- 2 * min(p_val_right, p_val_left)
  return(p_val)
}

p_val_links(mks_geno, N_permu=10000)
```


# Section 03 - Controlling for a 3rd variable
1. We found that marker 5211 and marker 5091 are associated with growth. However, we also found that
both markers are associated with each other. Thus, the association of one of these markers with growth could
be explained away by the association of the other one with growth.
Now, we would like to know if marker 5091 still associates with growth in maltose (YPMalt) when conditioned
on marker 5211. Define a null hypothesis, a statistic and use permutation testing to answer the question.
Strengthen your answer with a relevant plot.

marker 5091 no longer associates with growth, when conditioned by marker 5211

```{r}
growth_mrk <- merge(growth, genotype[marker == "mrk_5091", .(strain, genotype, marker)], 
                    by = 'strain', allow.cartesian = TRUE)
growth_mrk <- merge(growth_mrk, genotype[marker == "mrk_5211", .(strain, genotype, marker)], 
                    by = 'strain', suffixes = c("_5091", "_5211"), allow.cartesian = TRUE)
growth_mrk <- growth_mrk[media=="YPMalt"]
growth_mrk

facet_names <- list(
  'Lab strain'="Lab strain 5211",
  'Wild isolate'="Wild isolate 5211"
)
facet_labeller <- function(variable,value){
  facet_names[value]
}

ggplot(growth_mrk, aes(genotype_5091, growth_rate)) +
  geom_boxplot() +
  facet_wrap(~genotype_5211, labeller = facet_labeller) 

p_val_medians(growth_mrk[genotype_5211=="Lab strain", .(growth_rate, genotype=genotype_5091)] )
p_val_medians(growth_mrk[genotype_5211=="Wild isolate", .(growth_rate, genotype=genotype_5091)] )
```

2. Now, test if marker 5211 associates with growth in maltose when conditioned on marker 5091. Are the
results the same? Discuss.

marker 5211 association with growth persists, even when conditioned on marker 5091

```{r}
growth_mrk <- merge(growth, genotype[marker == "mrk_5211", .(strain, genotype, marker)], 
                    by = 'strain', allow.cartesian = TRUE)
growth_mrk <- merge(growth_mrk, genotype[marker == "mrk_5091", .(strain, genotype, marker)], 
                    by = 'strain', suffixes = c("_5211", "_5091"), allow.cartesian = TRUE)
growth_mrk <- growth_mrk[media=="YPMalt"]
growth_mrk

facet_names <- list(
  'Lab strain'="Lab strain 5091",
  'Wild isolate'="Wild isolate 5091"
)
facet_labeller <- function(variable,value){
  facet_names[value]
}

ggplot(growth_mrk, aes(genotype_5211, growth_rate)) +
  geom_boxplot() +
  facet_wrap(~genotype_5091, labeller = facet_labeller) 

p_val_medians(growth_mrk[genotype_5091=="Lab strain", .(growth_rate, genotype=genotype_5211)] )
p_val_medians(growth_mrk[genotype_5091=="Wild isolate", .(growth_rate, genotype=genotype_5211)] )
```




